src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -287,8 +287,8 @@ bool CanTokenFollowTypeInPattern(Precedence precedence)
                     ref openParenToken,
                     SyntaxKind.CloseParenToken,
                     static @this => @this.IsPossibleSubpatternElement(),
-                    static @this => @this.ParseSubpatternElement(),
-                    SkipBadPatternListTokens,
+                    @this => @this.ParseSubpatternElement(exitOnColon),
```

Comment:
`LanguageParser` has a `_termState` which includes many flags about what we are currently parsing, including `TerminatorState.IsSwitchSectionStart`. Is that something we could make use of here to simplify things? Tagging @CyrusNajmabadi in case he can weigh in.

Reply:
yes.  that seems better.  my bad :)

Reply:
If that's not actually an appropriate thing to use, we might want to consider somehow tracking this in the instance state of the parser. "What we are currently parsing" as we recurse through potentially many things seems like a reasonable thing to track, especially if it lets us potentially avoid a bunch of capturing delegate allocations.

Reply:
> What we are currently parsing" as we recurse through potentially many things seems like a reasonable thing to track,

Nope.  Def don't do that.  Will impact incremental parsing.  More parsing state is generally undesirable.

Reply:
OK, sounds like: let's try using `_termState`, and if something about that won't work, then we'll try to move forward with the general approach currently given in the PR.

Reply:
Switch section is the only child node nested in switch statement. It contains all case labels and statements ([example](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzguATEBqAPigdygQGMALAAgAoB7YAKwEoBYAKAG82LuKTUBTCgEYQXHnxSCATKNY95FYACd+EANYBuNgF8NQA=)). Since we can have a pattern inside a statement that comes after the case label, this flag cannot be used. Morover, it is not set when parsing pattern after `case` keyword, verified that it in the debugger. I can add a new flag, that will indicate, that we are parsing pattern in a case label of a switch statement. This will allow to not cary the boolean all over the place (which makes diff less and generally helps readability) and avoid closure captures. But it will introduce more state to the parser in such case. Which approache looks less evil?)

Reply:
you can add a specific termstate flag for the case-clause of a switch statement.  then enter/exit that term-state accordingly.  I'm totally fine with that.